lib/Mir.ml

```diff
@@ -65,49 +63,39 @@ and ('e, 's) statement =
   (* An SList does not share any of Block's semantics - it is just multiple
      (ordered!) statements*)
   | SList of 's list
-  | Decl of {decl_adtype: autodifftype; decl_id: string; decl_type: unsizedtype}
+  | Decl of
+      { decl_adtype: autodifftype
+      ; decl_id: string
+      ; decl_type: 'e sizedtype }
   | FunDef of
       { fdrt: unsizedtype option
       ; fdname: string
       ; fdargs: fun_arg_decl
       ; fdbody: 's }
-[@@deriving sexp, hash]
+[@@deriving sexp, hash, map]
 
-(** A "top var" is a global variable visible to the I/O of Stan.
-   Local vs. Global vardecls
-   There are "local" (i.e. not top-level; not read in or written out anywhere) variable
-   declarations that do not allow transformations. These are the only kind allowed in
-   the model block, and any declarations in a Block will also be local.
-   There are also then top-level ones, which are the only thing you can
-   write in both the parameters and data block. The generated quantities block allows both
-   types of variable declarations and, worse, mixes in top-level ones with normal ones.
-   We'll need to scan the list of declarations for top-level ones and essentially remove them
-   from the block. The AST has an `is_global` flag that also tracks this.
-*)
-type 'e top_var_decl =
-  { tvident: string
-  ; tvtype: 'e Ast.sizedtype
-  ; tvtrans: 'e Ast.transformation
-  ; tvloc: Ast.location_span sexp_opaque [@compare.ignore] }
-[@@deriving sexp]
+type io_block =
+  | Data
+  | Parameters
+  | TransformedParameters
+  | GeneratedQuantities
+[@@deriving sexp, hash]
 
-type 'e top_var_table = (string, 'e top_var_decl) Map.Poly.t [@@deriving sexp]
+type 'e io_var = string * ('e sizedtype * io_block) [@@deriving sexp]
 
 type ('e, 's) prog =
   { functions_block: 's list
-  ; data_vars: 'e top_var_table
-  ; tdata_vars: 'e top_var_table
-  ; prepare_data: 's list
-  ; params: 'e top_var_table
-  ; tparams: 'e top_var_table
+  ; input_vars: 'e io_var list
+  ; prepare_data: 's list (* data & transformed data decls and statements *)
   ; prepare_params:
       's list
-      (* XXX too intimately tied up with stan reader.hpp and writer.hpp in codegen
-     TODO: codegen parameter constraining and unconstraining in prepare_params
-  *)
-  ; log_prob: 's list
-  ; gen_quant_vars: 'e top_var_table
-  ; generate_quantities: 's list
+      (* param & transformed param decls and statements *)
+  ; log_prob: 's list (* model block, assuming above blocks run in scope*)
```

Comment:
Not sure what this comment means

Reply:
log_prob code assumes that data and params are already in scope and ready, maybe that's a better way to say it. This happens by running prepare_params and prepare_data first.

Reply:
Right! Maybe put that in the code comment?